Update libgraphics to account for Skia API changes

livecodeali opened the pull request:
The biggest impact is the removal of custom blend modes in Skia. This
forces us to drop support for legacy blend modes as we can no longer
emulate them.

On the plus side, only using the built-in Skia blend modes will result
in faster blending.

(cherry picked from commit 2ffc8ef900168f5ce0d791b9d6b72d3273a50dc6)

livecodefraser committed:
Update libgraphics to account for Skia API changes

The biggest impact is the removal of custom blend modes in Skia. This
forces us to drop support for legacy blend modes as we can no longer
emulate them.

On the plus side, only using the built-in Skia blend modes will result
in faster blending.

(cherry picked from commit 2ffc8ef900168f5ce0d791b9d6b72d3273a50dc6)

livecodeali committed:
[[ Skia ]] Ensure overloaded virtual warnings do not error when silenced

livecodeali committed:
[[ Travis CI ]] Build using gcc 4.9 on linux, clang on mac

livecodeali committed:
[[ Skia ]] Fix Mac desktop defines

livecodeian commented on libgraphics/src/context.cpp:
```diff
@@ -1350,34 +1463,42 @@ void MCGContextEnd(MCGContextRef self)
+		sk_sp<SkImage> t_image = t_child_layer->m_surface->makeImageSnapshot();
+		self->layer->canvas->save();
+		self->layer->canvas->resetMatrix();
+		self->layer->canvas->drawImage(t_image, t_child_layer->origin_x, t_child_layer->origin_y, &t_paint);
```
Could this be used instead? http://skia-doc.commondatastorage.googleapis.com/doxygen/doxygen/html/classSkSurface.html#a788d7e9a9eb7e93aef3dce9f684ea09f

livecodefraser replied:
Unfortunately, the resetMatrix bit is required - I did try it that way before and it caused issues.

livecodeian replied:
The resetMatrix stuff is fine, but the docs for SkSurface::draw implies that it could be more efficient than snapshot followed by drawImage

livecodeian commented on libgraphics/src/context.cpp:
```diff
@@ -2127,8 +2246,8 @@ static bool MCGContextApplyPaintSettingsToSkPaint(MCGContextRef self, MCGColor p
 		{
-			t_stipple = new (nothrow) SkStippleMaskFilter();
-			t_success = t_stipple != NULL;			
+			//t_stipple = new (nothrow) SkStippleMaskFilter();
```
If the stipple filter is no longer available then I suggest removing this code along with kMCPaintStyleStippled

livecodeian commented on libgraphics/src/w32text.cpp:
```diff
@@ -117,7 +119,7 @@ static bool w32_draw_text_process_transparent_buffer(uint32_t p_fill_color, bool
 			// and flag for opacity. Otherwise; take an average of the three RGB
 			// components and blend with the fill color.
-			if (((*t_context_pxl >> 24) & 0xFF) == 0xFF)
+			if (false && ((*t_context_pxl >> 24) & 0xFF) == 0xFF)
```
Should this dead code be removed?

livecodefraser replied:
The dead code gets removed in the update to use DirectWrite instead of GDI text rendering.

livecodeian commented on libgraphics/src/w32text.cpp:
```diff
@@ -249,25 +251,24 @@ static bool w32_draw_text_using_mask_to_context_at_device_location(MCGContextRef
 	if (t_success)
 	{
-		if ((*t_context_pxls >> 24) != 0xff)
+		if (true || (*t_context_pxls >> 24) != 0xff)
```
Can this if be removed now?

livecodeian commented on engine/src/mblandroidtypeface.cpp:
```diff
@@ -87,10 +90,12 @@ void MCAndroidTypefaceRelease(MCAndroidTypefaceRef p_typeface)
-	
+    // Skia APIs expect typefaces to be passed as shared pointers
+    sk_sp<SkTypeface> t_typeface((SkTypeface*)p_typeface);
+    t_typeface->ref();
```
looks like there's a convenience function for this - sk_ref_sp(T* obj) which wraps the ptr and increments the ref count

livecodeali committed:
[[ Skia ]] Update docs to remove legacy inks and gradients

livecodeali committed:
Merge branch 'develop' of github.com:livecode/livecode into feature-skia_update

Conflicts:
	libgraphics/src/legacyblendmodes.cpp
	libgraphics/src/legacygradients.cpp

livecodeali committed:
[[ Skia Update ]] Tweak a couple of things
- Use more efficient SkSurface::draw function
- Use sk_ref_sp convenience function when reffing sk_sp<SkTypeface>
  variables

livecodeian commented:
@livecode-vulcan review ok db257d5

livecode-vulcan commented:
💙 review by @livecodeian ok db257d5bbdbea99a186e8d768634e5a151ebe5a1
